File: Memorize/CardViewerTableViewController.m
//
//  CardViewerTableViewController.m
//  Memorize
//
//  Created by Heather Shelley on 4/20/14.
//  Copyright (c) 2014 Mine. All rights reserved.
//

#import "CardViewerTableViewController.h"
#import "UserDataController.h"
#import "Card.h"
#import "CardViewerViewController.h"

@interface CardViewerTableViewController ()

@property (nonatomic, strong) NSArray *nodes;

@end

@implementation CardViewerTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.tableView.allowsMultipleSelectionDuringEditing = NO;
    switch (self.viewType) {
        case ViewTypeReviewing:
            self.nodes = [[UserDataController sharedController] reviewingCards];
            break;
        case ViewTypeNotReviewing:
            self.nodes = [[UserDataController sharedController] notReviewingCards];
            break;
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.tableView reloadData];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSLog(@"count: %@", @(self.nodes.count));
    return self.nodes.count;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        Card *card = self.nodes[indexPath.row];
        [[UserDataController sharedController] deleteCard:card];
        NSMutableArray *mutableNodes = [self.nodes mutableCopy];
        [mutableNodes removeObject:card];
        self.nodes = mutableNodes;
        
        [self.tableView beginUpdates];
        [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
        [self.tableView endUpdates];
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell"];
    Card *card = self.nodes[indexPath.row];
    cell.textLabel.text = card.question;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [self performSegueWithIdentifier:@"ViewCardFromTable" sender:self.nodes[indexPath.row]];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([[segue identifier] isEqualToString:@"ViewCardFromTable"]) {
        [[segue destinationViewController] setExistingCard:sender];
    }
}

@end

File: Memorize/CardViewerTableViewController.swift
//
//  CardViewerTableViewController.swift
//  Memorize
//
//  Created by Heather Shelley on 2/25/17.
//  Copyright © 2017 Mine. All rights reserved.
//

import UIKit

enum ViewType {
    case reviewing
    case notReviewing
}

class CardViewerTableViewController: UITableViewController {
    var viewType = ViewType.reviewing
    var cards = [Card]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.allowsMultipleSelectionDuringEditing = false
        switch viewType {
        case .reviewing:
            cards = UserDataController.shared().reviewingCards() as? [Card] ?? []
        case .notReviewing:
            cards = UserDataController.shared().notReviewingCards() as? [Card] ?? []
        }
        
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        tableView.reloadData()
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        print("count: \(cards.count)")
        return cards.count
    }
    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return true
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        guard editingStyle == .delete else { return }
        
        let card = cards[indexPath.row]
        UserDataController.shared().delete(card)
        if let index = cards.index(of: card) {
            cards.remove(at: index)
            self.tableView.beginUpdates()
            self.tableView.deleteRows(at: [indexPath], with: .automatic)
            self.tableView.endUpdates()
        }
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        let card = self.cards[indexPath.row]
        cell.textLabel?.text = card.question
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        self.performSegue(withIdentifier: "ViewCardFromTable", sender: cards[indexPath.row])
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard segue.identifier == "ViewCardFromTable", let card = sender as? Card, let vc = segue.destination as? CardViewerViewController else { return }
        
        vc.existingCard = card
    }
}
